### PR TITLE
Add glob support to podman run/create --mount

### PR DESCRIPTION
**Pull request by rhatdan**
HPC Community asked for this support specifically for using GPUs
within containers. Nvidia requires the correct shared library to
to be present in the directory that matches the device mounted
into the container. These libraries have random suffixes based
on versions of the installed libraries on the host.

podman run --mount type=glob:src=/usr/lib64/nvidia\*:ro=true. This helps
quadlets be more portable for this use case.

#### Does this PR introduce a user-facing change?

```release-note
Podman --mount option supports bind mounts passed as globs.
```

### CODE REVIEW

File: docs/source/markdown/options/volume.md [hunk 1]
```diff
@@ -6,12 +6,16 @@
...
+not create the host path and will return an error. File globbing is supported,
+--volume /HOST-DIR/foo\* as long as destination paths are not specified.
+
+Similarly, `-v SOURCE-VOLUME:/CONTAINER-DIR` mounts the named volume from the
+host into the container. If no such named volume exists,
 Podman creates one. If no source is given, the volume is created
 as an anonymously named volume with a randomly generated name, and is
 removed when the <<container|pod>> is removed via the `--rm` flag or
-the `podman rm --volumes` command.
+the `podman rm --volumes` command. If the file path glob matches no sources, it will also be treated as an anonymous named volume.
```
Comment (mheon) [inline docs/source/markdown/options/volume.md:18]: Does this make sense? Honestly, I'd rather we just mount nothing in this case.

File: docs/source/markdown/options/volume.md [hunk 2]
```diff
@@ -6,12 +6,16 @@
 
 Create a bind mount. If `-v /HOST-DIR:/CONTAINER-DIR` is specified, Podman
 bind mounts `/HOST-DIR` from the host into `/CONTAINER-DIR` in the Podman
-container. Similarly, `-v SOURCE-VOLUME:/CONTAINER-DIR` mounts the named
-volume from the host into the container. If no such named volume exists,
+container. When specifying a host path, the host path must exist. Podman will
+not create the host path and will return an error. File globbing is supported,
+--volume /HOST-DIR/foo\* as long as destination paths are not specified.
```
Comment (mheon) [inline docs/source/markdown/options/volume.md:11]: I feel like this needs its own paragraph, it's sufficiently different from the rest (no destination supported) that it feels like it needs to be distinct.

File: pkg/specgen/volumes.go [hunk 3]
```diff
@@ -55,149 +55,202 @@ type ImageVolume struct {
...
-		// Support relative paths beginning with ./
-		if strings.HasPrefix(src, "./") {
-			path, err := filepath.EvalSymlinks(src)
-			if err != nil {
-				return nil, nil, nil, err
-			}
-			src, err = filepath.Abs(path)
-			if err != nil {
-				return nil, nil, nil, err
+	srcs, err := filepath.Glob(splitVol[0])
```
Comment (mheon) [inline pkg/specgen/volumes.go:76]: I know we've had CVEs in the past when we supported this with `podman cp`. I'm a little worried about allowing globs blindly.

Comment (mheon) [reply pkg/specgen/volumes.go:76]: Also, isn't `*` a valid character to have in a filename? Is that going to be a problem, or will glob expansion handle that without a problem?

File: pkg/specgen/volumes.go [hunk 4]
```diff
@@ -55,149 +55,202 @@ type ImageVolume struct {
...
+		switch {
+		case len(splitVol) == 1:
+			break
+		case len(splitVol) == 2:
+			if options, err = parse.ValidateVolumeOpts(strings.Split(splitVol[1], ",")); err != nil {
+				return nil, nil, err
 			}
-			splitVol[0] = src
+		default:
+			return nil, nil, errors.New("specifying globs with a desktination is not supported")
```
Comment (mheon) [inline pkg/specgen/volumes.go:95]: Typo destination

File: docs/source/markdown/options/volume.md (hunk 1 above)
Comment (rhatdan) [reply docs/source/markdown/options/volume.md:18]: The problem I have is this would be a breaking change.  In current, Poman
``` 
$ podman run -v /tmp/foo\* alpine ls -l /tmp
total 0
drwxr-xr-x    2 root     root             6 Jul 21 18:36 foo*
```

File: pkg/specgen/volumes.go (hunk 3 above)
Comment (rhatdan) [reply pkg/specgen/volumes.go:76]: ls /tmp/foo\* will match /tmp/foo\*

Comment (rhatdan) [reply pkg/specgen/volumes.go:76]: I am not sure if there is a risk of filepath.Glob(), it will basically just mount every file that matches the glob onto itself
inside of the container. If you had a sympolic link from one of the globs, I don't think filepath.Glob will resolve it, but when you end up with -v /tmp/foobar:/tmp/foobar which could be a symbolic link, but this should not be an issue.

File: docs/source/markdown/options/volume.md (hunk 1 above)
Comment (mheon) [reply docs/source/markdown/options/volume.md]: Still seems very confusing. I'll think about this over the weekend, but mounting a named volume in this case is absolutely not what I would expect.

File: docs/source/markdown/options/volume.md [hunk 5]
```diff
@@ -6,12 +6,20 @@
...
 Create a bind mount. If `-v /HOST-DIR:/CONTAINER-DIR` is specified, Podman
 bind mounts `/HOST-DIR` from the host into `/CONTAINER-DIR` in the Podman
-container. Similarly, `-v SOURCE-VOLUME:/CONTAINER-DIR` mounts the named
-volume from the host into the container. If no such named volume exists,
+container. When specifying a host path, the host path must exist. Podman will
+not create the host path and will return an error.
+
+File globbing is supported, --volume /HOST-DIR/foo\*. Not destination paths are
+not allowed. Paths matching the globs, will be mounted on the identical
+destination path within the container.
```
Comment (vrothberg) [inline docs/source/markdown/options/volume.md:14]:
```suggestion
File globbing is supported for sources (e.g., --volume /HOST-DIR/foo\*). 
Paths matching the globs, will be mounted on the identical
destination path within the container.
```

File: docs/source/markdown/options/volume.md [hunk 6]
```diff
@@ -6,12 +6,20 @@
...
+
+Similarly, `-v SOURCE-VOLUME:/CONTAINER-DIR` mounts the named volume from the
+host into the container. If no such named volume exists,
 Podman creates one. If no source is given, the volume is created
 as an anonymously named volume with a randomly generated name, and is
 removed when the <<container|pod>> is removed via the `--rm` flag or
-the `podman rm --volumes` command.
+the `podman rm --volumes` command. When the file path glob matches no paths on
+the host, the file path flob is treated as an anonymous named volume.
```
Comment (vrothberg) [inline docs/source/markdown/options/volume.md:22]:
```suggestion
the host, the file path glob is treated as an anonymous named volume.
```

Not sure why.  I would expect Podman to error out when the glob doesn't match.  Could be a user error which will be hard to debug when an anonymous volume is created instead.

File: test/system/160-volumes.bats [hunk 7]
```diff
@@ -558,4 +558,36 @@ EOF
...
 
 
+@test "podman volume globs" {
+    vol1a=${PODMAN_TMPDIR}/v1_$(random_string)
+    vol1b=${PODMAN_TMPDIR}/v1_$(random_string)
+    vol2=${PODMAN_TMPDIR}/v2_$(random_string)
+    touch $vol1a $vol1b $vol2
+
+    # if volumes source and dest match then pass
+    run_podman run --rm -v ${PODMAN_TMPDIR}/v1\*:ro $IMAGE ls $vol1a $vol1b
```
Comment (vrothberg) [inline test/system/160-volumes.bats:568]: What happens when I specify a source glob _and_ a destination?

**Review by Luap99 — COMMENTED** [review]
This is extremely confusing and backwards incompatible.
`-v /path` create anonymous volume at `/path` in container
`-v /path*` on the other hand means bind mount that path (expanding glob) from the host in the container, except when it 

--volume is already super overloaded with the syntax so I don't think we should do that.

I much rather see this as new option instead if we really need this

**Comment by vrothberg** [comment]
> I much rather see this as new option instead if we really need this

I like this idea.  The docs will be much easier to brain parse.

File: test/system/160-volumes.bats [hunk 8]
```diff
@@ -558,4 +558,36 @@ EOF
...
+    run_podman create --rm -v ${PODMAN_TMPDIR}/v1\*:ro $IMAGE ls $vol1a $vol1b
+    cid=$output
+    run_podman container inspect $output
+    is "$output" ".*$vol1a" "podman images --inspect should include $vol1a"
+    is "$output" ".*$vol1b" "podman images --inspect should include $vol1b"
+
+    run_podman 125 create --rm -v ${PODMAN_TMPDIR}/v3\*:ro $IMAGE echo fail
+    is "$output" "Error: invalid container path \"ro\", must be an absolute path" "podman create should fail since no glob match"
+
+    skip_if_remote "remote does not support CONTAINERS_CONF*"
```
Comment (edsantiago) [inline test/system/160-volumes.bats:580]: Please don't `skip` in the middle of a test: this will produce a log entry that does not reflect the fact that some tests ran and passed. Instead, please try:
```bash
    if ! is_remote; then
        ...indent the rest
```

**Review by edsantiago — CHANGES_REQUESTED** [review]

File: test/system/160-volumes.bats (hunk 7 above)
Comment (rhatdan) [reply test/system/160-volumes.bats:568]: It will fail.

**Comment by rhatdan** [comment]
I do not agree that this is super confusing, more of a corner case. BTW I don't like -v /foo creating an anonymous directory, I would have thought it as a short hand of -v /foo:/foo, so I found the existing issue confusing.  I think /foo versus /foo* is looking for a corner case, which is unlikely to confuse real users.  

We already have --mount  and --volume.  I don't think we should add a --glob or -vglob or some other option to make it harder to discover the option.

If you want to force this on --mount, I guess I would be ok with that.  But I would prefer this just enhance --volume.

**Comment by Luap99** [comment]
But is a breaking change. We cannot know how it is used today and we claim docker compatibility so this will hurt us in the long run. Corner cases are hard to document, hard to understand for users and hard to debug for us. We already have enough issues so I don't want to add problems that could have been easily avoided. 

Using `--mount` sounds possible, we could add a new type like `type=glob` for that.

**Comment by edsantiago** [comment]
I've been trying to find fault with this, because you know how I am, and I can't.  I can't think of any situation in which a volume name would contain asterisk, question mark, or bracket. And even if there are, for the first two, this code is a superset: `foo*` and `foo?` (actual asterisk and question mark in on-disk file) will still get mounted with `-v "foo*"/"foo?"` (asterisk/question mark in command line). I'm not seeing much possibility for confusion here.

I also tested with spaces, colons, semicolons, and everything behaved as it should.

Please squash commits though.

File: docs/source/markdown/options/volume.md (hunk 1 above)
Comment (rhatdan) [reply docs/source/markdown/options/volume.md]: Well there is no sure way to figure this out.  We could return an error if the user specifies a path that contains any 
"*", "?", "[" characters and the filepath.Glob did not match?  But this would prevent users from using volumes with these fileds in their name for a named volume.

**Comment by Luap99** [comment]
> I've been trying to find fault with this, because you know how I am, and I can't. I can't think of any situation in which a volume name would contain asterisk, question mark, or bracket. And even if there are, for the first two, this code is a superset: foo* and foo? (actual asterisk and question mark in on-disk file) will still get mounted with -v "foo*"/"foo?" (asterisk/question mark in command line). I'm not seeing much possibility for confusion here.


This is not what the current code does, a single path means create a anonymous volume and mount it at path in container.
The path is never used on the host. This PR complete changes that meaning if there is a glob character all of the sudden it means bind mount the path at the same place into the container.

```
# with this PR's code:
$ bin/podman run -v '/b*' -d --name c3  alpine top
# this creates volumes like this
{
                    "Type": "bind",
                    "Source": "/boot",
                    "Destination": "/boot",
                    "Driver": "",
                    "Mode": "",
                    "Options": [
                         "rbind"
                    ],
                    "RW": true,
                    "Propagation": "rprivate"
               },
               {
                    "Type": "bind",
                    "Source": "/bin",
                    "Destination": "/bin",
                    "Driver": "",
                    "Mode": "",
                    "Options": [
                         "rbind"
                    ],
                    "RW": true,
                    "Propagation": "rprivate"
               }

# now with this code when there is no match
$ bin/podman run -v '/z*' -d --name c3  alpine top
{
                    "Type": "volume",
                    "Name": "d057b9602f5675e58343d086604cb94d2358e685e12946188ba35d21ee17d042",
                    "Source": "/home/pholzing/.local/share/containers/storage/volumes/d057b9602f5675e58343d086604cb94d2358e685e12946188ba35d21ee17d042/_data",
                    "Destination": "/z*",
                    "Driver": "local",
                    "Mode": "",
                    "Options": [
                         "nosuid",
                         "nodev",
                         "rbind"
                    ],
                    "RW": true,
                    "Propagation": "rprivate"
               }
```

How is that not anything but confusing.

And then the big problem that this will break any user who uses these glob chars, linux allows all chars except NULL in the path so there can be users who use that. We must stay backwards compatible.
Note that the golang glob API also matches `[]`, saying people do not use these chars is just incorrect.



This is just a small example on my laptop, there are clearly applications using such characters.
```
$ find $HOME -name '*\]*' 2>/dev/null 
/home/pholzing/.local/share/containers/storage/overlay/3b33e9bd1fc97cc11f568c5fe7e8e1ecc78ffe206fee45da10ca9493852cc017/diff/usr/share/fonts/google-noto-vf/NotoSans-Italic[wght].ttf
/home/pholzing/.local/share/containers/storage/overlay/3b33e9bd1fc97cc11f568c5fe7e8e1ecc78ffe206fee45da10ca9493852cc017/diff/usr/share/fonts/google-noto-vf/NotoSans[wght].ttf
/home/pholzing/.debug/[kernel.kallsyms]
/home/pholzing/.debug/[vdso]
/home/pholzing/.thunderbird/b9aviz50.default-release/ImapMail/imap.gmail.com/[Gmail].msf
/home/pholzing/.thunderbird/b9aviz50.default-release/ImapMail/imap.gmail.com/[Gmail].sbd

$ find $HOME -name '*\?*' 2>/dev/null 
/home/pholzing/.cache/gnome-software/screenshots/624x351/71563fc1b6039a3b13d54bc1da929c756c5896a196ee5db72746d3f6f47194f5-FileRoller?action=AttachFile&do=get&target=file-roller.png
/home/pholzing/.cache/gnome-software/screenshots/112x63/71563fc1b6039a3b13d54bc1da929c756c5896a196ee5db72746d3f6f47194f5-FileRoller?action=AttachFile&do=get&target=file-roller.png
/home/pholzing/.cache/gnome-software/screenshots/752x423/71563fc1b6039a3b13d54bc1da929c756c5896a196ee5db72746d3f6f47194f5-FileRoller?action=AttachFile&do=get&target=file-roller.png

```

And then this here is just straight up broken:
```
$ bin/podman run -v '/e[' alpine true
Error: syntax error in pattern
```

**Comment by rhatdan** [comment]
Mainly because your example is a totally made up strange contrived example, that would never happen.  But if you want to argue it for the sake of winning an argument, you are right.

**Comment by mheon** [comment]
Would it be more sensible to require a destination path? `-v /var/lib/nvidia*:/var/lib:Z` remove potential anonymous-volume ambiguity and lets us alter the destination of the mounts if we want to.

**Comment by rhatdan** [comment]
How would that handle multiple matches.

**Comment by mheon** [comment]
Throw them all in the destination path. /var/lib/nvidia1 to $DESTDIR/nvidia1, /var/lib/nvidia2 to $DESTDIR/nvidia2

**Comment by rhatdan** [comment]
Paul will argue this is also a breaking change. since 
-v /var/lib/nvidia*:/var/lib:Z
Will currently take the /var/lib/nvidia\* and mount it on /var/lib.

**Comment by edsantiago** [comment]
I'm really liking Matt's destination-required idea. It adds what I think is an important flexibility that was missing from the original implementation:
```console
$ ... -v $HOME/test/volume*:/var/lib:Z ...
```
I still don't understand the problem with asterisk/questionmark, because this feature is a superset of current behavior (see my comment above). Brackets, I can see, but I wonder if this would solve it?
```go
    if _, err := Stat(splitVol[0]); err != nil {
        ...treat splitVol[0] as an actual true existing real thing
    } else {
        try globbing it...
```
I do think we should come up with a truth-table of scenarios, though, and run it through both current & proposed podman:
* `-v TMP/foo*:/tmp/foo`   where there are no files matching `fooanything`
* `-v TMP/foo*:/tmp/foo` where there are files or directories foo1, foo2
* `-v TMP/foo*:/tmp/foo` where there is an actual file called foo-star in addition to foo1 foo2
* .....

**Comment by mheon** [comment]
Maybe `-v /var/lib/nvidia*:/var/lib/*` with the trailing * on the destination meaning "in this directory" instead a literal mount to /var/lib?

**Comment by edsantiago** [comment]
First reaction: cringe, because seeing a star in a non-expansion context seriously gives me the heebie-jeebies. But I can breathe past that, because it's a valid idea. 

Second reaction: what does `/var/lib/nvidia*:/var/lib/*` mean? My eye finds it ambiguous:

| host | container, reasonable interpretation | container, mostly-but-not-entirely-unreasonable interpretation |
| ------ | ------ | ------ |
| /var/lib/nvidia1 | /var/lib/nvidia1 | /var/lib/1 |
| /var/lib/nvidia-foo | /var/lib/nvidia-foo | /var/lib/-foo |

Just a little bit of thinking convinces me that the right-hand column is silly, but a naïve reader seeing the above `-v` statement could reasonably be confused and have to spend time making sure that podman does the right thing. And then there's `-v /var/lib/nvidia*:/var/lib/nvidia/*` where I would genuinely feel baffled, because `/var/lib/nvidia/1` would be a reasonable interpretation of the right-hand wildcard. (Expectation-wise, not coding-wise).

I'm commenting in hopes of finding a good solution; this is not intended as a shoot-down.

**Comment by Luap99** [comment]
Of course my example is made up but the fact is that not allowing these charters is a breaking change if people try to mount such paths, currently we only block `:` which is already resulted in bug reports because people want to mount such paths. While I agree it is unlikely to use such chars in paths I still don't think that justifies the risk. And keep in mind some paths will just result in a random `syntax error in pattern` errors. 

And then I don't agree that a user can understand that
- `-v /path` means anonymous volume
- `-v /path*` means anonymous volume (assuming that glob does not match)
-  `-v /path*` means bind mount of `/path...` to `/path...` when the glob matches something

So that just does not make sense to me and does not match the documented syntax of `[[SOURCE-VOLUME|HOST-DIR:]CONTAINER-DIR[:OPTIONS]]`

As for using `-v /var/lib/nvidia*:/var/lib/*` I think a better alternative would be just `-v /var/lib/nvidia*:` where you leave the right side empty and when you need options then `-v /var/lib/nvidia*::ro`, this would result in better consistency as the left path is always on the host. And as the container path currently cannot be empty we can be 100% sure to not break anyone as this syntax was not allowed previously.


But I really do not see the advantage of trying to get this into --volume. Using --mount with a new type or a completely new option sounds much better to me. As for discoverability you can mention such option directly in the --volume paragraph. I don't see how this would be any worse.

**Comment by rhatdan** [comment]
I will move to the --mount method. Regrettably this is way less used, but at least for HPC community and anyone wanting to use devices or libraries, where the paths are not static, it makes it harder to deal with. 
The interesting point when we go to implement this in quadlet.  Currently Mount= does not exists, just Volume.

**Comment by vrothberg** [comment]
SGTM

I agree with the points Paul made.  The nice thing, however, in using volumes is that it's already being supported in containers.conf.  This means we need to consider adding a `Mounts = [""]` field (or sth similar) to containers.conf to meet the HPC needs.

**Comment by rhatdan** [comment]
@mheon @edsantiago @vrothberg @giuseppe @flouthoc @Luap99 PTAL

File: docs/source/markdown/options/mount.md [hunk 9]
```diff
@@ -26,10 +28,12 @@ Current supported mount TYPEs are **bind**, **volume**, **image**, **tmpfs** and
 
        Common Options:
 
-	      · src, source: mount source spec for bind and volume. Mandatory for bind.
+	      · src, source: mount source spec for bind, glob, and volume. Mandatory for bind and glob.
 
 	      · dst, destination, target: mount destination spec.
 
+	      Paths matching globs, are mounted on the destination directory with the identical name inside the container.
```
Comment (edsantiago) [inline docs/source/markdown/options/mount.md:35]: Extraneous comma, please remove. And, would it make sense to say "with the identical basename" or "identical filename"? Someone could possibly interpret that as "/foo/bar* will get mounted as /dest/foo/bar1, /dest/foo/bar2"

Comment (mheon) [inline docs/source/markdown/options/mount.md:35]: identical base name, maybe?

**Comment by mheon** [comment]
Code LGTM. Would appreciate more eyes on manpages though.

**Comment by rhatdan** [comment]
I triggered the merge, and can deal with the man page changes in a CI:DOC update.

**Comment by rhatdan** [comment]
Followup https://github.com/containers/podman/pull/19406
